Multi-batch message encoder

batiati opened the pull request:
## Multi-batch message encoder

This PR is the _reincarnation_ of #2617 and #2682: the same spirit in a brand new body.  
No code is reused from the previous attempts (except for parts of the encoder/decoder), as it takes a completely different approach.

Batching consists of the application submitting multiple independent units of work of the same operation (the batch payload) within a single VSR message.
This amortizes network and consensus costs, improving performance in scenarios where highly concurrent user requests submit operations containing only a few events each, sharing the same physical request.

This PR replaces the state machine batching mechanism based on the `Create{Accounts,Transfers}Result.index` field (aka Demuxer) with a general-purpose batching encoder/decoder suitable for any operation. (See #2617 for more details about the motivation).

### How it works:

The client (tb_client library) is responsible for accumulating concurrent requests, encoding them into multi-batch messages, and sending them as a single VSR request.

Upon receiving a multi-batch request, the StateMachine decodes the message body, taking advantage of prefetching the entire payload in a single pass whenever possible, while executing each batch as an independent unit of work.

New operation numbers for multi-batch guarantee backward compatibility with old clients sending non-encoded messages.

Each batch may generate a number of results which are encoded in a multi-batch reply and sent back to the client.

The client decodes the multi-batch reply's body into multiple batch payloads and calls the respective user callback for each one.

### Encoding layout:

- Multi-batch messages use a portion at the end of the message body (the batch trailer) to encode batch metadata, so batched requests can hold fewer events than non-batched ones.

- The trailer size is always a multiple of the event/result size to keep system invariants.

- The batch trailer consists of an array of `u16` values representing the number of elements in each batch, plus a _postamble_ `u16` containing the total number of batches encoded in the message.

- The trailer has a variable length, depending on the number of batches, but always in multiples of the event/result size.

- To facilitate batch encoding, the trailer is stacked from the end of the message towards the beginning. For example, the last value in the array of element counts corresponds to the first batch.

- Unused elements in the trailer, required for padding, are filled with `maxInt(u16)`.

For more details on the message encoding format, please see `multi_batch.zig` comments.

https://github.com/tigerbeetle/tigerbeetle/blob/1df370356e3ec783a5a2c5618c8cc1f9c1e8f6bf/src/vsr/multi_batch.zig#L24-L41

### This PR

See the commit messages for the implementation details.

- StateMachine support with unit tests.
- tb_client support with unit and echo tests.
- MultiBatch encoder/decoder fuzzer.
- VOPR multibatch and multi client release support.
- Vortex (java driver only) support for async calls, testing multibatch requests.
- Client integration tests for concurrent lookups and queries.

batiati committed:
[Docs] Update the new batch limit
Update the batch limit from 8190 to 8189 in docs, examples and client tests.

batiati committed:
[Clients] Integration tests
Add integration tests with concurrent lookups and queries.

batiati commented on src/vsr/replica.zig:
```diff
@@ -5510,7 +5516,9 @@ pub fn ReplicaType(
-            if (message.header.release.value > self.release.value) {
+            if (message.header.release.value > self.release.value and
+                !self.release.is_development())
+            {
```
I’m not particularly happy with this solution... but we need to test compatibility with clients <= `0.16.30` in a development cluster.

My first approach was to avoid relying on the `client_release` information for this, and instead create a new set of operations for multi-batch messages (`create_accounts_2`, `create_transfers_2`, etc.). But we already rely on `client_release` for other decisions (for example, `AccountFilterFormer`), so I think introducing a second mechanism for handling backwards compatibility would add unnecessary complexity right now.

Hopefully, we can deprecate all of this after a transition period (cutting the `minor = 17` release) and revisit this strategy after a proper code cleanup.

sentientwaffle replied:
:thinking: Yeah, this makes me nervous...

> I’m not particularly happy with this solution... but we need to test compatibility with clients <= 0.16.30 in a development cluster.

Which tests need this? 

Could we instead have `release.supports_multi_batching()`, and then have two different development releases `0.0.1`/`0.0.2` which return different values?

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -430,113 +473,159 @@ pub fn ContextType(
+                .pulse, .get_events => unreachable,
+                inline else => |operation_comptime| .{
+                    @sizeOf(StateMachine.EventType(operation_comptime)),
+                    @sizeOf(StateMachine.ResultType(operation_comptime)),
```
Earlier we use `event_size_bytes`/`result_size_bytes` -- should we be consistent or is there a reason to diverge here?

sentientwaffle commented on src/clients/c/tb_client/context.zig:
```diff
@@ -647,35 +736,34 @@ pub fn ContextType(
-                        }
-
+                        if (batched != packet_list) batched.assert_phase(.batched);
+                        it = batched.multi_batch_next;
```
Maybe pull this up into

```
while (it) |batched| : (it = batched.multi_batch_next) {
```

to make it more obvious? Right now it is hiding in between a couple of assertions.

sentientwaffle commented on src/vsr/multi_batch_fuzz.zig:
Also add this to `cfo.zig`'s task list!

sentientwaffle commented on src/stdx.zig:
```diff
@@ -173,6 +173,29 @@ pub fn zeroed(bytes: []const u8) bool {
+pub fn buffer_as_slice(comptime T: type, buffer: []u8) []T {
+    const size = @divFloor(buffer.len, @sizeOf(T)) * @sizeOf(T);
+    return @alignCast(std.mem.bytesAsSlice(T, buffer[0..size]));
+}
```
:thinking: Is this a common enough idiom to warrant promotion to stdx?
Also I don't think that the function name is descriptive enough -- from the name alone I would expect it to behave more like `bytesAsSlice`.

sentientwaffle commented on src/testing/cluster.zig:
```diff
@@ -297,7 +297,16 @@ pub fn ClusterType(comptime StateMachineType: anytype) type {
+                    const client_releases_supported = StateMachine.Workload.releases_supported;
+                    client.release = client_releases_supported[
+                        random.uintLessThan(usize, client_releases_supported.len)
+                    ];
```
This logic should move up a level, into `vopr.zig`.
Where practical, that is where the test parameter randomization takes place.

sentientwaffle commented on src/testing/vortex/java_driver/src/main/java/Main.java:
```diff
@@ -55,24 +68,42 @@ record Driver(Client client, Reader reader, Writer writer) {
-  void next() throws IOException {
+      * @throws ExecutionException
+      * @throws InterruptedException
+      */
```
Indentation?

sentientwaffle commented on src/testing/vortex/java_driver/src/main/java/Main.java:
```diff
@@ -113,6 +144,62 @@ void createAccounts(Reader reader, Writer writer, int count) throws IOException
     writer.flush();
   }
 
+  void createAccountsAsync(Reader reader, Writer writer, int count)
```
I might not be understanding this right, but does this function potentially take a batch from the user and split it into multiple physical requests?

sentientwaffle commented on docs/reference/requests/lookup_accounts.md:
Do we also need to change the warning in `query_accounts.md`?

sentientwaffle commented on src/tigerbeetle/benchmark_load.zig:
```diff
@@ -672,11 +719,46 @@ const Benchmark = struct {
+                    result,
+                    .{ .element_size = @sizeOf(tb.CreateAccountsResult) },
+                ) catch unreachable;
+                assert(reply_decoder.batch_count() == 1);
```
Could this part of the logic be pulled out of the switch and shared? e.g. via `@sizeOf(StateMachine.ResultType(...))`.

sentientwaffle reviewed (commented):
WIP review

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+        });
+        assert(buffer.len >= trailer_size_min);
+
+        // The end of the buffer must be aligned with the trailer.
+        // If it isn't, reduce the buffer to maintain alignment.
+        const aligned_len = std.mem.alignBackward(
+            usize,
+            buffer.len,
+            @sizeOf(TrailerItem),
+        );
```
Is the length ever not already a multiple of `@sizeOf(TrailerItem)`? It would be great if this could be an assert instead.

batiati replied:
No, the buffer size can be of any size (and we fuzz it with irregular sizes).
Only the final encoded message after calling `finish()` will be.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+            .batch_count = self.batch_count,
+            .element_size = self.options.element_size,
+        });
+        assert(buffer.len >= self.buffer_index + trailer_size);
```
I would swap the order of this clause so that it is more clear that it is a refinement of the previous assert.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+        maybe(self.batch_count == 0);
+
+        const element_count: u16 = element_count: {
+            if (self.options.element_size == 0) {
```
What about naming these `written_bytes` and `written_element_count`, so that it is clear that this is not the cumulative element count.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+
+        // For byte-aligned elements, padding may be required between the payload and the trailer.
+        const padding: u32 = self.buffer_index % @sizeOf(TrailerItem);
+        assert(padding < @sizeOf(TrailerItem));
```
Also to make this extra clear we can assert that either the padding is zero or the size of the element is one.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+        const postamble: *Postamble = @alignCast(@ptrCast(
+            buffer[self.buffer_index + padding + trailer_size - @sizeOf(Postamble) ..],
+        ));
+        postamble.batch_count = self.batch_count;
```
Instead set this as `postamble.*= ...` to make it clear that we are setting all fields.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+        // Supports zero-sized elements, or any power of two, including 2^0.
+        assert(options.element_size == 0 or std.math.isPowerOfTwo(options.element_size));
+
+        // Empty messages are considered valid.
```
I'm confused, where in the encoder do we create these empty messages?

batiati replied:
Not anymore, it was to simplify handling operations such `pulse` and `get_events` as always multi-batch.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,763 @@
+        return slice;
+    }
+
+    pub fn move_next(self: *MultiBatchDecoder) bool {
```
As far as I can tell this doesn't need to be public.

sentientwaffle commented on src/vsr/multi_batch_fuzz.zig:
```diff
@@ -0,0 +1,154 @@
+        batches.capacity(),
+    );
+
+    // Encoder will ignore and overwrite any existing content in the target buffer,
+    // at the end, both buffers must be equal.
```
Excellent idea!

sentientwaffle reviewed (commented):
WIP review

sentientwaffle commented on src/vsr/multi_batch_fuzz.zig:
```diff
@@ -0,0 +1,154 @@
+        usize,
+        1,
+        batches.capacity(),
+    );
```
Also add a randomized parameter to serve as another limit on `batch_element_count` -- as it is now, I think the vast majority of test runs will only use very few batches because it only takes a couple of large batches to use up all of the space.

batiati replied:
We do have `enum { zero, one, random }` for `batch_element_count`.

sentientwaffle commented on src/vsr/multi_batch_fuzz.zig:
Additionally fuzz the negative space, e.g.:

```diff
diff --git a/src/vsr/multi_batch_fuzz.zig b/src/vsr/multi_batch_fuzz.zig
index 71f33c958..e68379423 100644
--- a/src/vsr/multi_batch_fuzz.zig
+++ b/src/vsr/multi_batch_fuzz.zig
@@ -132,23 +132,52 @@ fn run_fuzz(options: struct {
     assert(encoder_bytes_written ==
         expect_payload_size + expect_padding + expect_trailer_size);
 
-    // Decode.
-    var decoder = try MultiBatchDecoder.init(
-        buffer_actual[0..encoder_bytes_written],
-        .{ .element_size = batch_element_size },
-    );
-    assert(expect_payload_size == decoder.payload.len);
+    {
+        // Decode.
+        var decoder = try MultiBatchDecoder.init(
+            buffer_actual[0..encoder_bytes_written],
+            .{ .element_size = batch_element_size },
+        );
+        assert(expect_payload_size == decoder.payload.len);
 
-    var payloads_decoded: u32 = 0;
-    for (batches.const_slice()) |batch_element_count| {
-        const batch_size: u32 = batch_element_count * batch_element_size;
-        const expect_batch: []u8 = buffer_expected[payloads_decoded..][0..batch_size];
-        payloads_decoded += batch_size;
+        var payloads_decoded: u32 = 0;
+        for (batches.const_slice()) |batch_element_count| {
+            const batch_size: u32 = batch_element_count * batch_element_size;
+            const expect_batch: []u8 = buffer_expected[payloads_decoded..][0..batch_size];
+            payloads_decoded += batch_size;
+
+            const decoded_batch = decoder.pop().?;
+            assert(batch_size == decoded_batch.len);
+            assert(std.mem.eql(u8, expect_batch, decoded_batch));
+        }
+        assert(payloads_decoded == decoder.payload.len);
+        assert(decoder.pop() == null);
+    }
 
-        const decoded_batch = decoder.pop().?;
-        assert(batch_size == decoded_batch.len);
-        assert(std.mem.eql(u8, expect_batch, decoded_batch));
+    // Verify that any flipped bit mutates the results or causes a decoding error
+    // (but never a panic).
+    for (0..32) |_| {
+        const byte_index = options.random.uintLessThan(usize, encoder_bytes_written);
+        for (0..@bitSizeOf(u8)) |bit_index| {
+            buffer_actual[byte_index] ^= @as(u8, 1) << @as(u3, @intCast(bit_index));
+            defer buffer_actual[byte_index] ^= @as(u8, 1) << @as(u3, @intCast(bit_index));
+
+            var decoder = MultiBatchDecoder.init(
+                buffer_actual[0..encoder_bytes_written],
+                .{ .element_size = batch_element_size },
+            ) catch continue;
+
+            var same: bool = true;
+            var payloads_decoded: u32 = 0;
+            for (batches.const_slice()) |batch_element_count| {
+                const batch_size: u32 = batch_element_count * batch_element_size;
+                const expect_batch: []u8 = buffer_expected[payloads_decoded..][0..batch_size];
+                payloads_decoded += batch_size;
+
+                const decoded_batch = decoder.pop().?;
+                same = same and std.mem.eql(u8, expect_batch, decoded_batch);
+            }
+            assert(!same);
+        }
     }
-    assert(payloads_decoded == decoder.payload.len);
-    assert(decoder.pop() == null);
 }
```

batiati committed:
[VSR] MultiBatch encoder and decoder
Implements the message encoder and decoder supporting the multi-batch format.

Although TigerBeetle supports batches as first-class primitives, sometimes applications need to submit independent units of work concurrently through the same client, creating the opportunity to pack them together into the same physical message to be executed within the same `op`.

MultiBatching solves the problem of how the client and the StateMachine should encode and decode these batches inside the message body by adding a thin trailer at the end of the message.
The trailer contains the number of batches encoded and the number of elements in each batch, and has a variable size, always a multiple of the operation's element size.

See the comments on `multi_batch.zig` for more details.

batiati committed:
[StateMachine] MultiBatch support
The state machine supports MultiBatch-encoded messages and replies, as well as legacy non-encoded messages for backwards compatibility.

Decoding and encoding are handled by the state machine (which is aware of multi-batching) and can optimize operations such as `prefetch` in a single pass for the entire payload (regardless of the number of batches encoded).
All operations support multi-batch encoding, even queries that require a single element per batch may have multiple filters encoded in the message body (bounded by `filter.limit`), handling operations where the max number of results does not depend on the number of events.

- Unit tests for both encoded and legacy messages.
- Unit tests for validate batches inputs.
- Benchmark and Repl

batiati committed:
[VOPR] MultiBatch support
Implements multi-batch encoder/decoder for VOPR, adding support for the Workload to pack multiple batches into a single request as well.
Adds support for testing legacy operations, stressing the logic for backwards compatibility.

batiati committed:
[tb_client] MultiBatch support
Code cleanup removing the client-side batching implementation (`Demuxer`) and helper functions such as `logical_batch_allowed` and linked chain validations.

Multi-batching is now handled by encoding/decoding messages, and all operations can be batched by the client, including queries.
Upgrade language clients bindings

batiati committed:
[Node] Remove dual buffer size
The node client used to allocate a single buffer during the request, large enough to store the packet and either the input or the reply buffer (whichever was larger). This approach allocated for the worst-case reply size (and always a full buffer of 8190 elements for queries).

Switch to a simpler approach where the buffer is allocated and freed between the request and reply with the exact required size. This adds more allocations, but optimizes memory usage in most cases.

batiati committed:
[Vortex] MultiBatch support

batiati committed:
[Docs] Update notice about batching.

batiati reviewed (commented):
The way we handle legacy operations was reworked after #2824.
It's ready for another review.

sentientwaffle commented on src/clients/c/test.zig:
```diff
@@ -133,16 +141,58 @@ test "tb_client echo" {
+        const event_size: u32, const event_request_max: u32 = switch (operation) {
+            .create_accounts => .{
+                @sizeOf(tb.Account),
+                @divExact(constants.message_body_size_max, @sizeOf(tb.Account)) - 1,
```
Is this -1 due to the trailer? If so it probably warrants a comment.

sentientwaffle commented on src/repl.zig:
```diff
@@ -824,24 +830,25 @@ pub fn ReplType(comptime MessageBus: type, comptime Time: type) type {
             operation: StateMachine.Operation,
-            arguments: []const u8,
+            buffer: []u8,
+            payload_size: u32,
```
Would it be simpler to handle the multi batching encoding at the caller site so that we don't have to pass in a partially used buffer + count? Or alternatively we could pass in the whole array list.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -554,39 +507,237 @@ pub fn StateMachineType(
+            }
+            assert(operation_is_multi_batch(operation));
+
+            const reply_trailer_size_min: u32 = vsr.multi_batch.trailer_total_size(.{
```
Would it be simpler if this function called `operation_result_max` for the reply part?

batiati replied:
Ideally, we would have only `operation_batch_max`, calculating the `min(event, result)`, but since queries are a special case, we need both `operation_event_max` and `operation_result_max`.
In the current implementation, it's `operation_result_max` that calls this function for batchable operations, as the number of results also depends on the number of events we can send.

https://github.com/tigerbeetle/tigerbeetle/blob/19b013670f1e7ee425709d99e25c69cecc14eeba/src/state_machine.zig#L646-L647

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -855,67 +1081,178 @@ pub fn StateMachineType(
                 .get_account_balances,
                 .query_accounts,
                 .query_transfers,
                 => |operation_comptime| {
```
Instead of a more verbose switch statement, would it make sense to just do a conditional based off of `if (operation_is_multi_batch(...)_) { ... } else { ... }`?

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -855,67 +1081,178 @@ pub fn StateMachineType(
+        /// Returns the logical time increment (in nanoseconds) for the highest
+        /// timestamp of the batch.
+        /// For multi-batch requests, this function expects a single, already decoded batch.
+        /// Inline function so that `operation` can be known at comptime.
```
I'm not sure what this comment means; I don't think this function can ever be called at compile time.

sentientwaffle commented on src/stdx.zig:
```diff
@@ -177,6 +177,83 @@ pub fn zeroed(bytes: []const u8) bool {
+/// always `@alignCast` the result.
+pub fn bytes_as_slice(
+    comptime T: type,
+    comptime mode: enum { exact, inexact },
```
Maybe swap the argument order to mirror `copy_disjoint`?

sentientwaffle commented on src/tigerbeetle/benchmark_load.zig:
```diff
@@ -367,18 +367,28 @@ const Benchmark = struct {
         assert(b.stage == .create_accounts);
 
-        const create_accounts_results = std.mem.bytesAsSlice(tb.CreateAccountsResult, result);
+        const create_accounts_results: []const tb.CreateAccountsResult = stdx.bytes_as_slice(
```
Why use an explicit type signature? Seems unnecessarily verbose.
(Likewise elsewhere -- I think the return value for `bytesAsSlice`/`bytes_as_slice` is well-known?)

sentientwaffle commented on src/tigerbeetle/benchmark_load.zig:
```diff
@@ -513,29 +542,45 @@ const Benchmark = struct {
+            tb.Account,
+            .exact,
+            result,
+        );
```
Why use `bytes_as_slice` (rather than `bytesAsSlice`) -- isn't `bytesAsSlice` actually _stricter_ since it doesn't cast alignment? (Also it is less verbose.)

batiati committed:
[Vortex] Fix operations.

sentientwaffle commented on src/state_machine/auditor.zig:
```diff
@@ -113,7 +113,7 @@ pub const AccountingAuditor = struct {
         client_count: usize,
 
-        batch_create_transfers_limit: u32,
+        transfers_expiry_max: u32,
```
"expiry max" makes this sound like a timestamp to me. What about `pulse_expiries_max`?

batiati committed:
[Fuzz] Remove global allocator.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,732 @@
+        /// The message body used, including the trailer.
+        body: []const u8,
+        options: Options,
+    ) error{MultiBatchInvalid}!MultiBatchDecoder {
```
What do you think of this refactor? My goal was to reduce the amount of manual offset calculations, and to make it super clear that checks like size/alignment are performed for each section.

```diff
diff --git i/src/vsr/multi_batch.zig w/src/vsr/multi_batch.zig
index 3ffba18d4..8a93c7202 100644
--- i/src/vsr/multi_batch.zig
+++ w/src/vsr/multi_batch.zig
@@ -121,53 +121,34 @@ pub const MultiBatchDecoder = struct {
         // Supports zero-sized elements, or any power of two, including 2^0.
         assert(options.element_size == 0 or std.math.isPowerOfTwo(options.element_size));
 
-        if (body.len < @sizeOf(Postamble)) return error.MultiBatchInvalid;
-        if (!std.mem.isAligned(
-            @intFromPtr(&body[body.len - @sizeOf(Postamble)]),
-            @alignOf(Postamble),
-        )) {
-            return error.MultiBatchInvalid;
-        }
+        const Parser = struct {
+            buffer: []const u8,
+            buffer_parsed: u32 = 0,
 
-        const postamble: *const Postamble = @alignCast(@ptrCast(
-            body[body.len - @sizeOf(Postamble) ..],
-        ));
-        if (postamble.batch_count == 0) return error.MultiBatchInvalid;
-        if (postamble.batch_count > multi_batch_count_max) return error.MultiBatchInvalid;
+            const Error = error{MultiBatchInvalid};
+            fn parse_suffix(parser: *@This(), comptime T: type, count: u32) Error![]const T {
+                const suffix_size = count * @sizeOf(T);
+                if (parser.buffer.len < suffix_size) return error.MultiBatchInvalid;
+                const suffix = parser.buffer[parser.buffer.len - suffix_size..];
+                const suffix_aligned = std.mem.isAligned(@intFromPtr(suffix.ptr), @alignOf(T));
+                if (!suffix_aligned) return error.MultiBatchInvalid;
+                parser.buffer = parser.buffer[0..parser.buffer.len - suffix_size];
+                parser.buffer_parsed += suffix_size;
+                return @alignCast(std.mem.bytesAsSlice(T, suffix));
+            }
+        };
 
+        var parser = Parser{ .buffer = body };
+        const postambles = try parser.parse_suffix(Postamble, 1);
+        const postamble = &postambles[0];
         const trailer_size = trailer_total_size(.{
             .element_size = options.element_size,
             .batch_count = postamble.batch_count,
         });
-        if ((options.element_size > 0 and body.len < trailer_size) or
-            (options.element_size == 0 and body.len != trailer_size))
-        {
-            return error.MultiBatchInvalid;
-        }
-        if (!std.mem.isAligned(
-            @intFromPtr(body[body.len - trailer_size ..].ptr),
-            @alignOf(TrailerItem),
-        )) {
-            return error.MultiBatchInvalid;
-        }
+        const trailer_items_used = try parser.parse_suffix(TrailerItem, postamble.batch_count);
+        const trailer_items_padding =
+            try parser.parse_suffix(u8, trailer_size - parser.buffer_parsed);
 
-        const trailer_items: []const TrailerItem = @alignCast(std.mem.bytesAsSlice(
-            TrailerItem,
-            body[body.len - trailer_size .. body.len - @sizeOf(Postamble)],
-        ));
-        if (trailer_items.len < postamble.batch_count) return error.MultiBatchInvalid;
-
-        const trailer_items_used = trailer_items[trailer_items.len - postamble.batch_count ..];
-        assert(trailer_items_used.len == postamble.batch_count);
-        if (trailer_items.len > postamble.batch_count) {
-            // Check the padding sentinel of the extra slots used for alignment.
-            const BackingInteger = @typeInfo(TrailerItem).Struct.backing_integer.?;
-            if (!std.mem.allEqual(
-                BackingInteger,
-                @ptrCast(trailer_items[0 .. trailer_items.len - postamble.batch_count]),
-                @bitCast(TrailerItem.padding),
-            )) return error.MultiBatchInvalid;
-        }
         const events_count_total: u32 = count: {
             var count: u32 = 0;
             for (trailer_items_used) |trailer_item| {
@@ -176,27 +157,28 @@ pub const MultiBatchDecoder = struct {
             break :count count;
         };
         if (options.element_size == 0 and events_count_total != 0) return error.MultiBatchInvalid;
-        const payload_size: u32 = std.math.mul(
-            u32,
-            events_count_total,
-            options.element_size,
-        ) catch |err| switch (err) {
+
+        const payload_size_or_overflow =
+            std.math.mul(u32, events_count_total, options.element_size);
+        const payload_size: u32 = payload_size_or_overflow catch |err| switch (err) {
             error.Overflow => return error.MultiBatchInvalid,
         };
 
         // For byte-aligned elements, padding may be required between the payload and the trailer.
-        const padding: u32 = @intCast(payload_size % @sizeOf(TrailerItem));
-        assert(padding < @sizeOf(TrailerItem));
-        assert(padding == 0 or options.element_size == 1);
-        if (!std.mem.allEqual(
-            u8,
-            body[body.len - padding - trailer_size ..][0..padding],
-            std.math.maxInt(u8),
-        )) return error.MultiBatchInvalid;
-        if (body.len != payload_size + padding + trailer_size) return error.MultiBatchInvalid;
+        const trailer_padding_size: u32 = @intCast(payload_size % @sizeOf(TrailerItem));
+        assert(trailer_padding_size < @sizeOf(TrailerItem));
+        assert(trailer_padding_size == 0 or options.element_size == 1);
+        const trailer_padding = try parser.parse_suffix(u8, trailer_padding_size);
+
+        for ([_][]const u8{ trailer_items_padding, trailer_padding }) |padding| {
+            if (!std.mem.allEqual(u8, padding, std.math.maxInt(u8))) return error.MultiBatchInvalid;
+        }
+
+        if (payload_size != body.len - parser.buffer_parsed) return error.MultiBatchInvalid;
+        assert(payload_size == parser.buffer.len);
 
         return .{
-            .payload = body[0..payload_size],
+            .payload = parser.buffer,
             .trailer_items = trailer_items_used,
             .payload_index = 0,
             .batch_index = 0,
```

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -3263,10 +4296,19 @@ pub fn StateMachineType(
 
-            const batch_create_accounts = operation_batch_max(.create_accounts, batch_size_limit);
-            const batch_create_transfers = operation_batch_max(.create_transfers, batch_size_limit);
+            const batch_create_accounts: u32 = @max(
+                operation_event_max(.create_accounts, batch_size_limit),
+                operation_event_max(.deprecated_create_accounts, batch_size_limit),
+            );
             assert(batch_create_accounts > 0);
+            assert(batch_create_accounts <= constants.batch_max.create_accounts);
+
+            const batch_create_transfers: u32 = @max(
+                operation_event_max(.create_transfers, batch_size_limit),
+                operation_event_max(.deprecated_create_transfers, batch_size_limit),
+            );
```
Aren't these the same values as the `batch_max` constants?

batiati replied:
No, it's based on `batch_size_limit` argument, which is the same in

```zig
const tree_values_count_max = tree_values_count(config.message_body_size_max);
```
but not in

```zig
const tree_values_count_limit = tree_values_count(options.batch_size_limit);
```

Actually, I've considered removing the `batch_max` constants completely, but they still have use to justify.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -752,13 +933,18 @@ pub fn StateMachineType(
+        scan_lookup_results: stdx.BoundedArrayType(
+            u32,
+            vsr.multi_batch.multi_batch_count_max,
+        ) = .{},
```
I think this is a 256KiB allocation, which is much more than we actually need. Instead of `multi_batch_count_max` could this be a `@max` of operation batch max's?

Also we should move this to the heap, since even at its minimal size this is a bit large to be on the stack.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,732 @@
+/// The maximum number of batches that can be encoded in a message.
+/// Since the multi-batch trailer has a variable length, the actual limit depends on the buffer
+/// size and the number of elements in each batch.
+pub const multi_batch_count_max: u16 = std.math.maxInt(u16);
```
Could this be `@min(maxInt(u16), message_body_size_max)`?

sentientwaffle reviewed (commented):
Review is still WIP, but posting what I have so far!

batiati committed:
Review

batiati committed:
Review stdx.bytes_as_slice

batiati committed:
Review: reduce the amount of manual offset calculations.

batiati committed:
Review: Repl

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -50,6 +50,9 @@ const constants = @import("../constants.zig");
 const vsr = @import("../vsr.zig");
 
 const Postamble = packed struct(u16) {
```
I assume that the reason that we aren't going to reserve any space in the postamble is that if want to (e.g.) add flags in the future then we can just create new operations, is that right?

batiati replied:
It was only natural that `Postamble` and `TrailerItem` had the same size (u16). To be honest, I haven't considered reserving extra space for future usage (like flags or encoding version).
Unless we can see practical use for it, I think we should keep operations as the primary mechanism for backward compatibility.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+
+        if (self.batch_index == self.trailer_items.len) {
+            assert(self.payload_index == self.payload.len);
+            return false;
```
I think this branch is unreachable. We only call this function from `pop` and it has already checked this condition.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+        const postamble: *const Postamble = postamble: {
+            const slice = try parser.parse_suffix(Postamble, 1);
+            break :postamble &slice[0];
+        };
```
```suggestion
        };
        if (postamble.batch_count == 0) return error.MultiBatchInvalid;
```
(Otherwise `trailer_total_size` can panic.)

This was caught by the [negative space fuzzing](https://github.com/tigerbeetle/tigerbeetle/pull/2787/files#r1996363952) -- could we get those into this branch?

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+        // element corresponds to the first batch.
+        const trailer_item: *const TrailerItem =
+            &self.trailer_items[self.trailer_items.len - self.batch_index - 1];
+        maybe(trailer_item.element_count == 0);
```
Since we immediately check this condition in an `if` statement, the `maybe` statement is redundant.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+
+        const batch_item: []const u8 = self.peek();
+        const moved = self.move_next();
+        maybe(moved);
```
We can remove `move_next` entirely and replace it with:
```suggestion
        self.batch_index += 1;
        self.payload_index += @intCast(batch_item.len);
```

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+        if (self.batch_count == 0) {
+            assert(self.buffer_index == 0);
+            return 0;
+        }
```
It seems like we are being inconsistent over whether it is legal to encode a multi batch with no batches. The encoder appears to allow it, but the documentation says it isn't allowed, and the decoder fails. e.g.:

```
test "batch: encode no batches" {
    const buffer_size = (1024 * 1024) - @sizeOf(vsr.Header); // 1MiB message.
    const buffer = try testing.allocator.alignedAlloc(u8, @alignOf(vsr.Header), buffer_size);
    var encoder = MultiBatchEncoder.init(buffer, .{ .element_size = 128 });
    const bytes_written = encoder.finish();

    var decoder = MultiBatchDecoder.init(
        buffer[0..bytes_written],
        .{ .element_size = 128 },
    ) catch unreachable;
    assert(decoder.batch_count() == 0);
}
```

batiati replied:
That's true! We removed support here https://github.com/tigerbeetle/tigerbeetle/pull/2787#discussion_r1994336961, so it's illegal to encode zero batches (but we allow zero-element batches)

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+        element_size: u32,
+        buffer: []align(16) u8,
+        batch_count: u16,
+        elements_per_batch: ?u16 = null,
```
Or `batch_elements`.

sentientwaffle commented on src/vsr/multi_batch.zig:
```diff
@@ -0,0 +1,745 @@
+        });
+
+        // Cleaning the buffer first, so it can assert the bytes.
+        @memset(options.buffer, std.math.maxInt(u8));
```
Deleting this line doesn't change the test results. What assert is this referring to? If we don't need a particular value then we could randomize the buffer instead, or memset undefined.

batiati committed:
StateMachineFuzz: multi batch support.

batiati committed:
MultiBatchFuzz:fuzz negative space

batiati committed:
Review

sentientwaffle commented on src/vsr/multi_batch_fuzz.zig:
```diff
@@ -0,0 +1,181 @@
+    // Verify that any flipped bit mutates the results or causes a decoding error
+    // (but never a panic).
+    for (0..32) |_| {
+        const byte_index = options.prng.int_inclusive(usize, encoder_bytes_written);
```
Since the rng is inclusive, this is off-by-one:
```suggestion
        const byte_index = options.prng.int_inclusive(usize, encoder_bytes_written - 1);
```

sentientwaffle commented on src/vsr/multi_batch_fuzz.zig:
```diff
@@ -0,0 +1,181 @@
+
+                const decoded_batch = decoder.pop().?;
+                same = same and std.mem.eql(u8, expect_batch, decoded_batch);
+            }
```
```suggestion
            }
            assert(decoder.pop() == null);
```

batiati committed:
Review
